### PR TITLE
Auth: Registration with Certicate signed request

### PR DESCRIPTION
**Pull request by eloycoto**
This commit adds new parameters to the "registration" directive, where a user
will send a CSR to the endpoint, the operator will ask CA provider to get it
signed and return the certificate to the user.

At the moment, CSR is not required, and the main reason is to not break the
clients right now, but a follow-up commit will fix that and it'll implement
Authz and the missing tests.

Signed-off-by: Eloy Coto <eloy.coto@acalustra.com>

### CODE REVIEW

File: internal/yggdrasil/yggdrasil.go [hunk 1]
```diff
@@ -304,22 +307,49 @@ func (h *Handler) PostDataMessageForDevice(ctx context.Context, params yggdrasil
...
 		logger.V(1).Info("received registration info", "content", registrationInfo)
+
+		res := models.MessageResponse{
+			Directive: msg.Directive,
+			MessageID: msg.MessageID,
+		}
+		content := models.RegistrationResponse{}
+
+		_, err = h.deviceRepository.Read(ctx, deviceID, h.initialNamespace)
+		if err == nil {
```
Comment (jakub-dzon) [inline internal/yggdrasil/yggdrasil.go:326]: Shouldn't it be `err != nil`? `err == nil` means that the device has already been registered and in the past we ignored such registration request. From now on is that a way to re-issue a certificate?

File: internal/yggdrasil/yggdrasil_test.go [hunk 2]
```diff
@@ -56,26 +65,52 @@ var _ = Describe("Yggdrasil", func() {
...
 		configMap          *configmaps.MockConfigMap
 
 		errorNotFound = errors.NewNotFound(schema.GroupResource{Group: "", Resource: "notfound"}, "notfound")
 		boolTrue      = true
+
+		k8sClient client.Client
+		testEnv   *envtest.Environment
 	)
 
+	initKubeConfig := func() {
```
Comment (jakub-dzon) [inline internal/yggdrasil/yggdrasil_test.go:79]: Please change the name of the test file to `yggdrasil_integration_test`

File: internal/yggdrasil/yggdrasil.go (hunk 1 above)
Comment (eloycoto) [reply internal/yggdrasil/yggdrasil.go:326]: Certificate  expired, and we need to renew it. device worker already started to get 401 status code

**Review by pkliczewski — COMMENTED** [review]
Do you think it makes sense to add some tests?

**Comment by eloycoto** [comment]
> Do you think it makes sense to add some tests?

What I added has full test coverture and I'll add integration test too :-)

**Review by masayag — APPROVED** [review]

**Pull request merged**